### PR TITLE
 connectivity: Add connectivity tests for multiple policy types

### PR DESCRIPTION
**Pull request by sayboras**
### Description

This PR is to add coverage for below policy tests:

- named port
- ICMP policy
- policy with matched expression
- policy with service account rule
- denied policies

### CODE REVIEW

File: connectivity/manifests/client-egress-to-cidr-1111-deny.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# This policy allows packets towards 1.1.1.1, but not 1.0.0.1.
+# Both addresses are owned by CloudFlare/APNIC.
```
Comment (aanm) [inline connectivity/manifests/client-egress-to-cidr-1111-deny.yaml:2]: Can you explain this a little bit more? From my local testing all traffic is being denied.

Comment (aanm) [reply connectivity/manifests/client-egress-to-cidr-1111-deny.yaml:2]: Also, it looks like this file is not used anywhere.

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
@joestringer @jrajahalme can you also take a 2nd pair of :eyes: into this PR just to make sure the results of the tests are the ones that we should expect?

File: connectivity/manifests/client-egress-to-cidr-1111-deny.yaml (hunk 1 above)
Comment (sayboras) [reply connectivity/manifests/client-egress-to-cidr-1111-deny.yaml:2]: good pick, I supposed to adjust this file, and raise one issue for deny with except policy.

Comment (sayboras) [reply connectivity/manifests/client-egress-to-cidr-1111-deny.yaml]: here is the github issue https://github.com/cilium/cilium/issues/20914, I have reduced the scope of this test here.

Comment (sayboras) [reply connectivity/manifests/client-egress-to-cidr-1111-deny.yaml]: Seems like this is expected behavior, I will update docs as suggested in above github issue.

File: connectivity/manifests/client-egress-to-echo-expression-deny.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    toEndpoints:
+    - matchLabels:
+        io.kubernetes.pod.namespace: kube-system
+        k8s-app: kube-dns
+    - matchLabels:
+        io.kubernetes.pod.namespace: kube-system
+        k8s-app: coredns
+    - matchLabels:
+        io.kubernetes.pod.namespace: kube-system
+        k8s-app: node-local-dns
```
Comment (aanm) [inline connectivity/manifests/client-egress-to-echo-expression-deny.yaml:32]: This is a good opportunity to test `matchExpressions`, something like (untested):
```yaml
    - matchLabels:
        io.kubernetes.pod.namespace: kube-system
      matchExpressions:
      - key: "k8s-app"
        op: "equals"
        values:
         - "kube-dns"
         - "coredns"
         - "node-local-dns"
```

**Review by aanm — CHANGES_REQUESTED** [review]

File: connectivity/manifests/client-egress-to-echo-expression-deny.yaml (hunk 2 above)
Comment (sayboras) [reply connectivity/manifests/client-egress-to-echo-expression-deny.yaml:32]: good point, let me update as per the above.

I have another case testing with matchExpressions, but with exists function as well.

File: connectivity/suite.go [hunk 3]
```diff
@@ -240,6 +275,124 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+			WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+				if a.Destination().HasLabel("kind", "client") {
+					return check.ResultOK, check.ResultDrop
+				}
+				return check.ResultOK, check.ResultOK
+			})
+
+		// This policy allows ingress to echo only from client with a label 'other:client'
+		ct.NewTest("echo-ingress-deny").WithPolicy(echoIngressFromOtherClientDenyPolicyYAML).
+			WithScenarios(
```
Comment (joestringer) [inline connectivity/suite.go:318]: Maybe drop this comment about allowing 1.1.1.1 since it conflicts with the comments above and also the policy?

File: connectivity/suite.go [hunk 4]
```diff
@@ -240,6 +275,124 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+				// should lead to a drop.
+				if a.Destination().Path() == "/private" && !a.Destination().HasLabel("X-Very-Secret-Token", "42") {
+					return check.ResultDropCurlHTTPError, check.ResultNone
+				}
+				egress = check.ResultOK
+				// Expect all curls from client2 to be proxied and to be GET calls.
+				egress.HTTP = check.HTTP{
+					Method: "GET",
+				}
+				return egress, check.ResultNone
```
Comment (joestringer) [inline connectivity/suite.go:297]: nit: For consistency and easier review, it wolud be nice if this test name matched the policy name (`client-egress-to-echo-deny`)

File: connectivity/suite.go [hunk 5]
```diff
@@ -240,6 +275,124 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
 			return check.ResultDrop, check.ResultNone
 		})
```
Comment (joestringer) [inline connectivity/suite.go:277]: What's the difference in this test vs. the previous one? Both seem to be basic label selection for denies, it's only the content of the labels themselves which is different. This seems like it's already covered by `client-ingress-from-client2-deny`?

File: connectivity/suite.go (hunk 4 above)
Comment (joestringer) [reply connectivity/suite.go:297]: The above applies for a few of the tests in this commit, but I didn't bother to comment on each one.

File: connectivity/suite.go [hunk 6]
```diff
@@ -154,7 +184,6 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
 			return check.ResultOK, check.ResultDrop
 		})
 
-	// This policy allows ingress to echo only from client with a label 'other:client'.
```
Comment (joestringer) [inline connectivity/suite.go:157]: nit: Deleting this line seems unrelated to the rest of the changes? The comment is still valid right?

File: connectivity/suite.go [hunk 7]
```diff
@@ -240,6 +275,124 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+		ct.NewTest("client-egress-expression-deny").WithPolicy(clientEgressToEchoExpressionDenyPolicyYAML).
+			WithScenarios(
+				tests.PodToPod(),
+			).
+			WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+				if a.Destination().HasLabel("kind", "echo") &&
+					a.Source().HasLabel("kind", "client") {
+					return check.ResultDrop, check.ResultNone
+				}
+				return check.ResultOK, check.ResultOK
```
Comment (joestringer) [inline connectivity/suite.go:364]: I got a bit confused by this default allow, but I think pod-to-pod only ever sends traffic from client -> echo so all cases are covered in the condition above, right? So I guess this code is never hit.

File: connectivity/suite.go [hunk 8]
```diff
@@ -187,6 +216,12 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
 			tests.PodToPod(),
 		)
 
+	// This policy allows port 8080 from client to echo (using label match expression, so this should succeed
+	ct.NewTest("client-egress-expression").WithPolicy(clientEgressToEchoExpressionPolicyYAML).
+		WithScenarios(
+			tests.PodToPod(),
+		)
```
Comment (joestringer) [inline connectivity/suite.go:223]: Why are there no Expectations on this test case?

**Review by joestringer — CHANGES_REQUESTED** [review]

File: connectivity/check/policy.go [hunk 9]
```diff
@@ -26,6 +26,7 @@ const (
 	ExitAnyError    ExitCode = -1
 	ExitInvalidCode ExitCode = -2
 
+	ExistInvalidHost  ExitCode = 6
```
Comment (joestringer) [inline connectivity/check/policy.go:29]: Oh, and is this change intentional? I don't see this being used anywhere in the PR.

File: connectivity/suite.go (hunk 3 above)
Comment (joestringer) [reply connectivity/suite.go:318]: Sorry, I'm trying a new code review tool... this is referring to line 389 below.

File: connectivity/suite.go (hunk 4 above)
Comment (joestringer) [reply connectivity/suite.go:297]: ^ See line 305 and all teh other ones with `client-ingress-deny` as the test name.

File: connectivity/suite.go (hunk 5 above)
Comment (joestringer) [reply connectivity/suite.go:277]: ^ This was meant to comment on `clientIngressFromClient2DenyPolicyYAML` vs. `echoIngressFromOtherClientDenyPolicyYAML`

File: connectivity/check/policy.go (hunk 9 above)
Comment (jrajahalme) [reply connectivity/check/policy.go:29]: Is the `Exist` prefix here intended, maybe it should be `Exit`, i.e., `ExitInvalidHost`?

File: connectivity/manifests/client-egress-only-dns-deny.yaml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+apiVersion: cilium.io/v2
+kind: CiliumNetworkPolicy
+metadata:
+  namespace: cilium-test
+  name: client-egress-only-dns
```
Comment (jrajahalme) [inline connectivity/manifests/client-egress-only-dns-deny.yaml:5]: This name should also indicate that this is a deny policy.

File: connectivity/manifests/client-egress-to-cidr-deny.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+apiVersion: "cilium.io/v2"
+kind: CiliumNetworkPolicy
+metadata:
+  namespace: cilium-test
+  name: client-egress-to-cidr-deny
+spec:
+  endpointSelector:
+    matchLabels:
+      kind: client
+  egressDeny:
```
Comment (jrajahalme) [inline connectivity/manifests/client-egress-to-cidr-deny.yaml:10]: I don't now remember the semantics a policy with a sole deny rule should have... Isn't this denying everything, like if there was no rules in the policy? Should we add some allow rules to be able to make a distinction between deny/allow? Especially if a wider rule allows (e.g., everything to a entity), and a deny rule then denies some ports (that would be allowed based on the 1st rule)??

File: connectivity/suite.go [hunk 12]
```diff
@@ -240,6 +254,74 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
 			return check.ResultDrop, check.ResultNone
 		})
 
+	// Tests with deny policy
+	if ct.DenyPolicy() {
+		// This policy denies ingress into client from client2.
+		ct.NewTest("client-ingress-deny").WithPolicy(clientIngressFromClient2DenyPolicyYAML).
+			WithScenarios(
+				tests.ClientToClient(),
```
Comment (jrajahalme) [inline connectivity/suite.go:307]: Would be good to have traffic from non-clients as well to validate the policy on traffic that does not match the deny rule?

File: connectivity/suite.go [hunk 13]
```diff
@@ -240,6 +254,74 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+			).
+			WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+				if a.Destination().HasLabel("kind", "client") {
```
Comment (jrajahalme) [inline connectivity/suite.go:310]: It looks like we'd need to check the label of the source as well, especially if we add non-client sources like I suggested above.

File: connectivity/suite.go [hunk 14]
```diff
@@ -240,6 +254,74 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+
+		// This policy allows ingress to echo only from client with a label 'other:client'
+		ct.NewTest("echo-ingress-deny").WithPolicy(echoIngressFromOtherClientDenyPolicyYAML).
+			WithScenarios(
+				tests.PodToPod(),
+			).
+			WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+				if a.Destination().HasLabel("kind", "echo") {
+					return check.ResultDrop, check.ResultDrop
+				}
```
Comment (jrajahalme) [inline connectivity/suite.go:324]: Based on this echo pods would be dropping traffic regardless the source, while the policy seems to be specific to the "other client"?

Comment (jrajahalme) [reply connectivity/suite.go:324]: Also, is the drop happening already at the egress as the return values would seem to indicate. If so how come we get drop also from ingress?

File: connectivity/suite.go [hunk 15]
```diff
@@ -240,6 +254,74 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+				return check.ResultOK, check.ResultOK
+			})
+
+		// This policy denies ingress to echo only from client with a label 'other:client'.
+		ct.NewTest("client-ingress-icmp-deny").WithPolicy(echoIngressICMPDenyPolicyYAML).
+			WithFeatureRequirements(check.RequireFeatureEnabled(check.FeatureICMPPolicy)).
+			WithScenarios(
+				tests.ClientToClient(),
```
Comment (jrajahalme) [inline connectivity/suite.go:332]: Would be nice to also have non-ICMP traffic to validate the difference in policy behavior, if any?

File: connectivity/suite.go [hunk 16]
```diff
@@ -240,6 +254,74 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+		// This policy denies L3 traffic to 1.0.0.1/8 CIDR from client to echo
+		ct.NewTest("to-cidr-deny").
+			WithPolicy(clientEgressToCIDRDenyPolicyYAML).
+			WithScenarios(
+				tests.PodToCIDR(),
+			).
+			WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+				if a.Destination().Address() == "1.1.1.1" ||
+					a.Destination().Address() == "1.0.0.1" {
+					// Expect packets for 1.1.1.1 to be allowed.
```
Comment (jrajahalme) [inline connectivity/suite.go:389]: Comment needs updating.

File: connectivity/manifests/client-egress-l7-http-named-port.yaml [hunk 17]
```diff
@@ -0,0 +1,39 @@
+---
+# client2 is allowed to contact one.one.one.one/ on port http-80 and the echo Pod
```
Comment (jrajahalme) [inline connectivity/manifests/client-egress-l7-http-named-port.yaml:2]: CIDR/toFQDN destinations can't have named ports, so maybe remove that from the comment? Policy seems correct in this regard already.

File: connectivity/suite.go [hunk 18]
```diff
@@ -352,6 +398,34 @@ func Run(ctx context.Context, ct *check.ConnectivityTest) error {
...
+		WithScenarios(
+			tests.PodToPod(),
+			tests.PodToWorld(),
+		).
+		WithExpectations(func(a *check.Action) (egress, ingress check.Result) {
+			if a.Source().HasLabel("other", "client") && // Only client2 is allowed to make HTTP calls.
+				// Outbound HTTP to one.one.one.one is L7-introspected and allowed.
+				(a.Destination().Port() == 80 && a.Destination().Address() == "one.one.one.one" ||
+					a.Destination().Port() == 8080) { // 8080 is traffic to echo Pod.
+				if a.Destination().Path() == "/" || a.Destination().Path() == "" {
```
Comment (jrajahalme) [inline connectivity/suite.go:440]: would add to the end of the comment: `via named port`. Otherwise this validation code is kind of hard to read w.r.t. where a named port was used.

**Review by jrajahalme — CHANGES_REQUESTED** [review]
Cilium docs say about deny policies:
```
Similarly to "allow" policies, Pods will enter default-deny mode as soon a
single policy selects it.
```

Based on this each deny policy should also maybe allow some wider set of traffic to test that the deny actually takes precedence, and that the deployed datapath policy was not a default deny without any actual deny rules in it!

File: connectivity/manifests/client-egress-to-echo-expression-deny.yaml (hunk 2 above)
Comment (sayboras) [reply connectivity/manifests/client-egress-to-echo-expression-deny.yaml:32]: thanks, op should be `In`, then it's working as expected.

File: connectivity/suite.go (hunk 5 above)
Comment (sayboras) [reply connectivity/suite.go:277]: good catch, let me keep echoIngressFromOtherClientDenyPolicyYAML as this test will test same + cross node traffic.

File: connectivity/suite.go (hunk 7 above)
Comment (sayboras) [reply connectivity/suite.go:364]: yeah, I guess partially it was me, who didn't know about the default deny :man_facepalming: 

I will update the default behaviour to Drop for both Ingress and Egress.

File: connectivity/manifests/client-egress-to-cidr-deny.yaml (hunk 11 above)
Comment (sayboras) [reply connectivity/manifests/client-egress-to-cidr-deny.yaml:10]: this is good point, I didn't know about default deny till recently.

Each deny tests now are having one extra policy (allow all except world) and one positive test.

**Comment by sayboras** [comment]
Thanks all for your detailed review, I have added additional policy for deny tests, so that default deny can be avoided. Extra policies could be a little permissive, but should be good enough to verify behavior under test.

File: connectivity/suite.go (hunk 8 above)
Comment (joestringer) [reply connectivity/suite.go:232]: FWIW if you don't comment about this in the code, I'm pretty sure someone is going to look at it again in future and ask the exact same question. Omission is not always deliberate, so if you deliberately omit something then it's often easier to communicate by explicitly omitting it, eg by adding a line afterwards describing why this doesn't have any expectations.

**Review by aanm — APPROVED** [review]

File: connectivity/suite.go (hunk 6 above)
Comment (tklauser) [reply connectivity/suite.go:157]: This line is still deleted by this particular commit (37a0224bc257a5b34b329998198eeb400f6d5651).

File: internal/cli/cmd/connectivity.go [hunk 19]
```diff
@@ -119,6 +119,7 @@ func newCmdConnectivityTest() *cobra.Command {
 	cmd.Flags().StringVar(&params.AssumeCiliumVersion, "assume-cilium-version", "", "Assume Cilium version for connectivity tests")
 	cmd.Flags().BoolVarP(&params.Verbose, "verbose", "v", false, "Show informational messages and don't buffer any lines")
 	cmd.Flags().BoolVarP(&params.Debug, "debug", "d", false, "Show debug messages")
+	cmd.Flags().BoolVarP(&params.DenyPolicy, "deny-policy", "", true, "Run tests with deny policy")
```
Comment (tklauser) [inline internal/cli/cmd/connectivity.go:122]: Do we really need this specific flag? Couldn't we just use something like `'!/.*-deny' instead to skip deny policy tests?

File: connectivity/manifests/client-egress-to-echo.yaml [hunk 20]
```diff
@@ -21,12 +21,6 @@ spec:
...
-        k8s-app: kube-dns
-    - matchLabels:
-        io.kubernetes.pod.namespace: kube-system
-        k8s-app: coredns
-    - matchLabels:
-        io.kubernetes.pod.namespace: kube-system
-        k8s-app: node-local-dns
+    - matchExpressions:
+      - { key: 'k8s-app', operator: In, values: [ "kube-dns", "coredns", "node-local-dns" ] }
+      - { key: 'io.kubernetes.pod.namespace', operator: In, values: [ "kube-system" ] }
```
Comment (tklauser) [inline connectivity/manifests/client-egress-to-echo.yaml:26]: Nit: missing newline at end of file

**Comment by tklauser** [comment]
Seems like my review message got lost somehow, so repeating it here:

To me it looks like this PR is now introducing quite a bit more than just deny policy tests, e.g. also tests for named ports and matched expressions. I think this should at least be mentioned in the PR title (so it shows up in the release notes) and maybe also briefly summarized in the PR description rather than just the commit messages.

**Comment by sayboras** [comment]
> To me it looks like this PR is now introducing quite a bit more than just deny policy tests, e.g. also tests for named ports and matched expressions. I think this should at least be mentioned in the PR title (so it shows up in the release notes) and maybe also briefly summarized in the PR description rather than just the commit messages.

Good point 💯, let me update the PR description

File: internal/cli/cmd/connectivity.go (hunk 19 above)
Comment (sayboras) [reply internal/cli/cmd/connectivity.go:122]: do you know how to skip multiple patterns ? I tried a few cases but failed, so just add the flag for easier control 🤔

Comment (tklauser) [reply internal/cli/cmd/connectivity.go:122]: It looks like `--test` is [defined](https://github.com/cilium/cilium-cli/blob/374d13759ea397c8c28bd840b57254b46d1e577d/internal/cli/cmd/connectivity.go#L116) as a [`StringSliceVar`](https://pkg.go.dev/github.com/spf13/pflag#StringSliceVar), so either multiple invocations (i.e. `--test '!/*-deny' --test '!/pod-to-.*-nodeport'`) or a comma-separate `--test` invocation (i.e. `--test '!/*-deny,!/pod-to-.*-nodeport'`, like e.g. [here](https://github.com/cilium/cilium-cli/blob/374d13759ea397c8c28bd840b57254b46d1e577d/.github/in-cluster-test-scripts/eks-tunnel.sh#L32)) should work. If it doesn't, then we should fix/extend `--test` parsing/handling rather than introducing new flags for each special case.

**Review by tklauser — APPROVED** [review]
LGTM, thanks Tam!

**Comment by joestringer** [comment]
:wave: @sayboras I noticed that this only specifies the endpoint selector for which pod to apply the policy with the serviceaccount, but it doesn't use the serviceaccount in the ingress / egress policy statements. Was that intentional?

**Comment by sayboras** [comment]
> 👋 @sayboras I noticed that this only specifies the endpoint selector for which pod to apply the policy with the serviceaccount, but it doesn't use the serviceaccount in the ingress / egress policy statements. Was that intentional?

It's not :(, let me double-check and do a follow-up on this.

**Comment by sayboras** [comment]
This PR should cover the gap from the above comment 

https://github.com/cilium/cilium-cli/pull/1532
